File: +ndi/documentservice.m
classdef documentservice
% ndi.documentservice - a class of methods that allows objects to interact with ndi.document objects
%
    properties (SetAccess=protected, GetAccess=public)

    end; % properties

    methods
        function ndi_documentservice_obj = documentservice()
            %ndi.documentservice - create an ndi.documentservice object, which is just an abstract class
            %
            % NDI_DOCUMENTSERVICE_OBJ = ndi.documentservice();
            %
                
        end; % ndi.documentservice()

        function ndi_document_obj = newdocument(ndi_documentservice_obj)
            % NEWDOCUMENT - create a new ndi.document based on information in this object
            %
            % NDI_DOCUMENT_OBJ = NEWDOCUMENT(NDI_DOCUMENTSERVICE_OBJ)
            %
            % Create a new ndi.document based on information in this class.
            %
            % The base ndi.documentservice class returns empty.
            %
                ndi_document_obj = [];
        end; % newdocument

        function sq = searchquery(ndi_documentservice_obj)
            % SEARCHQUERY - create a search query to find this object as an ndi.document
            %
            % SQ = SEARCHQUERY(NDI_DOCUMENTSERVICE_OBJ)
            %
            % Return a search query that can be used to find this object's representation as an
            % ndi.document.
            %
            % The base class ndi.documentservice just returns empty.
                sq = [];
        end; % searchquery
    end; 
end

File: ndi-python/ndi/documentservice.py
"""
NDI DocumentService - Mixin for objects that can create and search documents.
"""

from typing import TYPE_CHECKING, Any, Dict, List, Optional

if TYPE_CHECKING:
    from .document import Document
    from .query import Query


class DocumentService:
    """
    Mixin class for objects that provide document services.

    This class provides an interface for creating new documents and
    searching for documents in a database.
    """

    def newdocument(self, document_type: str = 'base', **properties) -> 'Document':
        """
        Create a new document.

        Args:
            document_type: The type of document to create
            **properties: Property name/value pairs to set

        Returns:
            Document: A new document instance
        """
        raise NotImplementedError("Subclasses must implement newdocument()")

    def searchquery(self) -> 'Query':
        """
        Return a search query for documents associated with this object.

        Returns:
            Query: A query object
        """
        raise NotImplementedError("Subclasses must implement searchquery()")
